controller.go

```diff
@@ -904,6 +897,13 @@ addToStore:
 
 	c.arrangeUserFilterRule()
 
+	// It is finally safe to increment the Endpoint count for the --config-from parent network
+	if t != nil {
```

Comment:
This seems to be the key logic change.  Previously, the deferred function was guarded on `err == nil`.   In the case that `--config-from` networks exit ungracefully we arrive at the end of the function with `err != nil` from one of the many error checks, and this causes the `IncEndpointCnt()` call to be skipped.

I think you can leave this logic as a deferred function, and in that case you also don't have to move `t` into the enclosing function's scope.   However I'd like to understand better why `err != nil` at the end of the function, and whether there are any other consequences of unconditionally updating the reference count in this case - is there another `err != nil` case we need to think about?

Do the logs on the issues allow you to pinpoint which error check is failing?  I couldn't immediately find an error message which matched.

Reply:
Thanks for taking a look @euanh 
If a user creates a `--config-only` network with the incorrect parent interface e.g.
`docker network create -d macvlan --subnet 10.0.48.0/22 --gateway 10.0.48.1 -o parent=eth0-invalid --config-only 10-0-48`

and then creates a `--config-from` network and attach it to a service e.g.

`docker network create -d macvlan --scope swarm --config-from 10-0-48 --attachable private-10-0-48`
`docker service create --detach=false --network name=private-10-0-48,alias=${HOSTNAME} --replicas 1 --name bla nginx:alpine`

the code bails out at https://github.com/docker/libnetwork/blob/9ff9b57c344df5cd47443ad9e65702ec85c5aeb0/controller.go#L826

with an `invalid subinterface vlan name` error

This should have ideally not executed https://github.com/docker/libnetwork/blob/9ff9b57c344df5cd47443ad9e65702ec85c5aeb0/controller.go#L805
but it does and bumps up the endpoint refcount for the `--config-only` network and so it becomes impossible to delete the parent network

Reply:
Ok, the root of the problem is that the call to `getConfigNetwork` shadows `err` in the outer scope:

https://github.com/docker/libnetwork/blob/9ff9b57c344df5cd47443ad9e65702ec85c5aeb0/controller.go#L796-L797

The `defer` then closes over the _shadow_ `err`, which must always be `nil` otherwise the `defer` is never even evaluated:
https://github.com/docker/libnetwork/blob/9ff9b57c344df5cd47443ad9e65702ec85c5aeb0/controller.go#L804-L805

I presume the original intention was that the deferred function would be conditional on the _outer_ err, and so would only increment if no errors had occurred in the rest of the function.   In that case you just need to avoid capturing `err` in the `if` block:

```suggestion
        if network.configFrom != "" {
	    var t *network
	    t, err = c.getConfigNetwork(network.configFrom)
	    ...
```

Reply:
We might want to look at using a `retErr` return parameter, which we would set explicitly and test in the deferred function.   It seems to be a pattern for dealing with complicated functions which have to do a lot of error handling.

Reply:
@euanh I've updated the diff to make sure the function using the same `err` variable throughout its scope (except in the deferred functions)
I also moved the `var` block to the top of the function because `network` seems to be a type and a variable and was causing a conflict